pennylane_orquestra/orquestra_device.py

```diff
@@ -382,3 +387,254 @@ def qubit_operator_string(self, observable):
+
+        # 4-5. Create the backend specs & workflow file
+        workflow = gen_expval_workflow(
+            self.qe_component,
+            self.backend_specs,
+            qasm_circuit,
+            ops,
+            resources=self._resources,
+            **kwargs,
+        )
+        file_id = str(uuid.uuid4())
+        filename = f"expval-{file_id}.yaml"
+        filepath = write_workflow_file(filename, workflow)
+
+        # 6. Submit the workflow
+        workflow_id = qe_submit(filepath, keep_file=self._keep_files)
+
+        if self._keep_files:
+            self._filenames.append(filename)
+
+        self._latest_id = workflow_id
+
+        # 7. Loop until finished
+        results = self.single_step_results(workflow_id)
+
+        # Insert the theoretical value for the expectation value of the
+        # identity operator
+        for idx in identity_indices:
+            results.insert(idx, 1)
+
+        res = self._asarray(results)
+        return res
+
+    def single_step_results(self, workflow_id):
+        """Extracts the results of a single step obtained for a workflow.
+
+        This method assumes that the workflow had a single step and that the
+        structure of the result corresponds to results sent by Orquestra API
+        v1.0.0.
+
+        Args:
+            workflow_id (str): the ID of the workflow to extract results for
+
+        Returns:
+            results (list): a list of workflow results
+        """
+        data = loop_until_finished(workflow_id, timeout=self._timeout)
+        try:
+            step_result = [v for k, v in data.items()][0]
+            results = step_result["expval"]["list"]
+        except (IndexError, KeyError, TypeError, AttributeError) as e:
+            current_status = workflow_details(workflow_id)
+            raise ValueError(
+                f"Unexpected result format for workflow {workflow_id}.\n "
+                "{''.join(current_status)}"
+            )
+        return results
+
+    @staticmethod
+    def insert_identity_res_batch(results, empty_obs_list, identity_indices):
+        """An auxiliary function for inserting values which were not computed
+        using workflows into batch results.
+
+        Computations involving the identity observable are given by theoretical
+        values rather than as part of a workflow. Therefore, such values need
+        to be inserted into the results later.
+
+        Args:
+            results (list): workflow results of the batched execution
+            empty_obs_list (list): list of indices where every observable was
+                identity
+            identity_indices (dict): maps the index of a sublist to the
+                the list of indices where the observable was an identity
+
+        Returns:
+            list: list of results
+        """
+        # Insert the lists needed for only identity results
+        for idx in empty_obs_list:
+            results.insert(idx, [])
```

Comment:
Since we are mutating the `results` list as part of the loop, will this cause an inconsistency (e.g, will the correct `idx` start to 'drift'?)

Reply:
Hi @josh146, that's a great question! Actually was also wondering about this when adding it. One thing to note is that `empty_obs_list` is special in two ways: a) it is sorted in a strictly ascending order b) it stores indices with regards to the complete list (not a partial version).

In that sense, `idx` can be:
a) `idx==0` (inserting to the head of the list, but only once)
b) in-between existing elements (e.g., if `idx<len(results)`)
c) last (e.g., if `idx==len(results)`)

*Why is this fine for each case?*
a) and c) are fine, as we are always appending to either the head or the tail of the list using indices of the complete list.

b) Since `empty_obs_list` is sorted, we can be sure that for the next `idx` the sublist `results[:idx]` is already complete, as no insertion will happen to an `i<idx` index. So basically after insertion, `results[:idx+1]` will not be mutated.

The second half of the list, `results[idx+1:]` is still incomplete this time and elements in `results[idx+1:]` are not at their correct indices yet and can be "pushed back".

*Example*

Creating the `[[], [], [], 1, [], 2]` list
```python
In [8]: results = [1,2]  
In [10]: empty_obs_list = [0, 1,2, 4]
In [11]: for idx in empty_obs_list: 
    ...:       results.insert(idx, []) 
    ...:   
In [12]: results                                                                                                                                  
Out[12]: [[], [], [], 1, [], 2]
```

Reply:
Thanks Antal! Good explanation